### ISSUE TITLE
Exception raising does not narrow types.

### ISSUE THREAD

**Issue by montyevans** [microsoft/pyright#2804]
**Describe the bug**
If a variable X has a type that is the union of two types, A and B, and you raise an exception if X is type A, then any code after the exception is only reachable if X is type B. Pyright should probably be able to recognize this and narrow the type accordingly, but it doesn't.

**To Reproduce**
```python
x: Optional[int] = None
if x is None: raise ValueError("Variable x cannot be None")
x += 4
```
**Expected behavior**
The line `x += 4` should not raise any type checking errors.

**Screenshots or Code**
![image](https://user-images.githubusercontent.com/32969601/148120460-9e055777-6134-47a2-83b7-78f2935d7b59.png)

The issue occurs both with the VSCode extension and the command-line tool. I'm using pyright 1.1.203

**Comment by erictraut** [comment]
Pyright already does this. In your example above, you're assigning `None` to `x`, so it is already narrowed to `None` when it evaluates the expression `if x is None`. In the negative ("implied else") case, it is narrowed to `Never`. When pyright detects that it's narrowing `Never`, it backs off and doesn't perform narrowing. Instead, it reverts back to the declared type.

You can see it working as you expect if you eliminate the assignment.

```python
def func(x: Optional[int]):
    if x is None:
        raise ValueError("Variable x cannot be None")
    x += 4
```

**Comment by montyevans** [comment]
Ah, thanks Eric - silly of me to miss the earlier narorwing!
